src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,362 @@
+
+            var newSyntaxTree = RazorSyntaxTree.Create(rewrittenRoot, syntaxTree.Source, syntaxTree.Diagnostics, syntaxTree.Options);
+            return newSyntaxTree;
+        }
+
+        public static RazorSyntaxTree RemoveMarkupElement(RazorSyntaxTree syntaxTree)
+        {
+            var remover = new Remover();
+            var rewrittenRoot = remover.Visit(syntaxTree.Root);
+
+            var newSyntaxTree = RazorSyntaxTree.Create(rewrittenRoot, syntaxTree.Source, syntaxTree.Diagnostics, syntaxTree.Options);
+            return newSyntaxTree;
+        }
+
+        private class Rewriter : SyntaxRewriter
+        {
+            // From http://dev.w3.org/html5/spec/Overview.html#elements-0
+            private static readonly HashSet<string> VoidElements = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+            {
+                "area",
+                "base",
+                "br",
+                "col",
+                "command",
+                "embed",
+                "hr",
+                "img",
+                "input",
+                "keygen",
+                "link",
+                "meta",
+                "param",
+                "source",
+                "track",
+                "wbr"
+            };
+
+            private readonly Stack<TagBlockTracker> _startTagTracker;
+
+            public Rewriter()
+            {
+                _startTagTracker = new Stack<TagBlockTracker>();
+            }
+
+            private TagBlockTracker CurrentTracker => _startTagTracker.Count > 0 ? _startTagTracker.Peek() : null;
+
+            public string CurrentStartTagName => CurrentTracker?.TagName;
+
+            public override SyntaxNode Visit(SyntaxNode node)
+            {
+                if (node != null)
+                {
+                    node = RewriteNode(node);
+                }
+
+                return base.Visit(node);
+            }
+
+            private SyntaxNode RewriteNode(SyntaxNode node)
+            {
+                if (node.IsToken ||
+                    node is MarkupElementSyntax ||
+                    node.Parent is MarkupElementSyntax)
+                {
+                    // This is either a token or we have already rewritten this node.
+                    return node;
+                }
+
+                var currentIndex = 0;
+                _startTagTracker.Clear();
+                var children = node.ChildNodes();
+                while (currentIndex < children.Count)
+                {
+                    var child = (RazorSyntaxNode)children[currentIndex];
+                    if (!(child is MarkupTagBlockSyntax tagBlock) ||
+                        tagBlock.Parent is MarkupElementSyntax)
+                    {
+                        // Not a node that we want to rewrite.
+                        currentIndex++;
+                        continue;
+                    }
+
+                    var tagName = tagBlock.GetTagName();
```

Comment:
`GetTagName()` is an extension method to `MarkupTagBlockSyntax`. This will change in the future when tag block have a first class TagName child.